### DISCUSSION TITLE
What do the "checkout" and "checkin" mean in charon log?

### DISCUSSION THREAD

**Discussion by andy-ning** [strongswan/strongswan#2491] [Q&A]
I see many IKE_SA checkout and checkin messages in charon log, what do the "checkout" and "checkin" mean exactly? Are they indicating IKE_SAs are disconnected and re-connected? 

Thanks.

Some snippet of the logs:

 7840 24-10-07 12:13:09.312 09[MGR] checkout IKEv2 SA with SPIs 150cc55c34b6ed67_i a5e1bc4c957f4893_r
 7841 24-10-07 12:13:09.312 09[MGR] IKE_SA system-nodes[586] successfully checked out
 7842 24-10-07 12:13:09.312 09[KNL] <system-nodes|586> querying policy fdff:719a:bf60:4080::4/128 === fdff:719a:bf60:4080::2/128 in
 7843 24-10-07 12:13:09.312 09[MGR] <system-nodes|586> checkin IKE_SA system-nodes[586]
 7844 24-10-07 12:13:09.312 09[MGR] <system-nodes|586> checkin of IKE_SA successful

 7845 24-10-07 12:13:13.046 06[MGR] checkout IKEv2 SA with SPIs 1e99c150f50b6fca_i eef1e84d909b8533_r
 7846 24-10-07 12:13:13.046 06[MGR] IKE_SA system-nodes[587] successfully checked out
 7847 24-10-07 12:13:13.046 06[KNL] <system-nodes|587> querying policy fdff:719a:bf60:4080::4/128 === fdff:719a:bf60:4080::3/128 in
 7848 24-10-07 12:13:13.046 06[MGR] <system-nodes|587> checkin IKE_SA system-nodes[587]
 7849 24-10-07 12:13:13.046 06[MGR] <system-nodes|587> checkin of IKE_SA successful

 7850 24-10-07 12:13:16.047 07[MGR] checkout IKEv2 SA with SPIs 1e99c150f50b6fca_i eef1e84d909b8533_r
 7851 24-10-07 12:13:16.047 07[MGR] IKE_SA system-nodes[587] successfully checked out
 7852 24-10-07 12:13:16.047 07[KNL] <system-nodes|587> querying policy fdff:719a:bf60:4080::4/128 === fdff:719a:bf60:4080::3/128 in
 7853 24-10-07 12:13:16.047 07[KNL] <system-nodes|587> querying policy fdff:719a:bf60:4080::3/128 === fdff:719a:bf60:4080::4/128 out
 7854 24-10-07 12:13:16.047 07[MGR] <system-nodes|587> checkin IKE_SA system-nodes[587]
 7855 24-10-07 12:13:16.047 07[MGR] <system-nodes|587> checkin of IKE_SA successful

**Comment by tobiasbrunner** [comment]
> Are they indicating IKE_SAs are disconnected and re-connected?

No. The IKE daemon is multi-threaded and whenever a thread is working on an IKE_SA, whether it's for handling an incoming IKE message or some kind of local event, the IKE_SA is exclusively checked out from the IKE_SA manager, afterwards it's checked in again so other threads can work on it. There are various scheduled jobs that e.g. handle regular NAT keepalives or DPDs that check out an IKE_SA to query the usage stats on the attached IPsec SAs/policies and then send a message if necessary.

**Reply by andy-ning** [reply]
Thanks for the detailed explanation.
